File: leetcode/strings/valid-palindrome-two-pointers-from-center.js
// 2 pointers from the center moving outwards
const isValidPalindromeCenter = function(s) {
  s = s.replace(/[^A-Za-z0-9]/g, "").toLowerCase();

  // Initialize left/right pointers to point at the middle index
  // of the string. Indexes start at 0 meaning that we have to
  // floor() the value from dividing length by 2 in order to get
  // the index of the center.
  let left = Math.floor(s.length / 2), right = left;

  // If the string is even, more left pointer back by 1 so left/right
  // are pointing at the 2 middle values respectively.
  if (s.length % 2 === 0) {
    left--;
  } 

  // Loop throught the string wile expanding pointers outwards
  // comparing the characters.
  while (left >= 0 && right < s.length) {
    if (s[left] !== s[right]) {
      return false;
    }
    left--;
    right++;
  }

  return true;
}

try {
  Error.stackTraceLimit = 1;
  if (!(isValidPalindromeCenter("aabaa") === true)) throw new Error("Test Result Error");
  if (!(isValidPalindromeCenter("aabbaa") === true)) throw new Error("Test Result Error");
  if (!(isValidPalindromeCenter("abc") === false)) throw new Error("Test Result Error");
  if (!(isValidPalindromeCenter("a") === true)) throw new Error("Test Result Error");
  if (!(isValidPalindromeCenter("") === true)) throw new Error("Test Result Error");
  if (!(isValidPalindromeCenter(" ") === true)) throw new Error("Test Result Error");
  if (!(isValidPalindromeCenter("A man, a plan, a canal: Panama") === true)) throw new Error("Test Result Error");
  if (!(isValidPalindromeCenter("race a car") === false)) throw new Error("Test Result Error");
  if (!(isValidPalindromeCenter("race car") === true)) throw new Error("Test Result Error");
  console.log("All test cases passed!");
}
catch(err) {
  console.log(err.stack)
}

File: leetcode/strings/valid_palindrome_two_pointers_from_center.rb
# https://leetcode.com/problems/valid-palindrome/
# Given a string s, return true if it is a palindrome, or false otherwise.
# A phrase is a palindrome if, after converting all uppercase letters into
# lowercase letters and removing all non-alphanumeric characters, it reads
# the same forward and backward.

# Two pointers from center:
# @param {String} s
# @return {Boolean}
def is_palindrome(s)
  s = s.gsub(/[^A-Za-z0-9]/, "").downcase
  
  length = s.length
  return true if length <= 1
  
  left = length / 2
  right = left
  left -= 1 if length.even?
  
  while (left >= 0 && right < length)
    if s[left] != s[right]
      return false
    end
    
    left -= 1
    right += 1
  end
  
  return true 
end

if __FILE__ == $0
  begin
    raise RuntimeError unless is_palindrome("aabaa") == true
    raise RuntimeError unless is_palindrome("aabbaa") == true
    raise RuntimeError unless is_palindrome("abc") == false
    raise RuntimeError unless is_palindrome("a") == true 
    raise RuntimeError unless is_palindrome("") == true
    raise RuntimeError unless is_palindrome(" ") == true
    raise RuntimeError unless is_palindrome("A man, a plan, a canal: Panama") == true
    raise RuntimeError unless is_palindrome("race a car") == false
    raise RuntimeError unless is_palindrome("race car") == true 
  rescue RuntimeError => e
    puts e.message + ": " + e.backtrace.inspect
  else
    puts "All test cases passed!"
  end    
end